### DISCUSSION TITLE
Pre-compiling Lua?

### DISCUSSION THREAD

**Discussion by sawcce** [mlua-rs/mlua#137] [Q&A] [answered]
Hi!
I was wondering if it was possible to pre-compile Lua code to for example LuaJIT bytecode or something of the sort. I'm trying to execute Lua code but I don't necessarily want people to be able to see the raw code in the final binaries.
I've searched all around the documentation but I can't seem to find anything about that...

Sawcce

**Comment by atweiden** [comment]
[How rlua does this](https://github.com/amethyst/rlua/issues/178)

https://github.com/amethyst/rlua/pull/236

**Comment by khvzak** [comment, marked as answer]
You can load chunk and then dump it to bytecode:
```rust
let bytecode = lua.load("return 1 + 1").into_function()?.dump(true);

// To load it back
let result = lua.load(&bytecode).set_mode(ChunkMode::Binary).eval::<i32>()?;
assert_eq!(result, 2);
```

Starting from 0.8.0 bytecode can be loaded without restrictions (since too many ways available to load bytecode and too much effort to protect all that places).
On the previous versions Lua must be created via `unsafe_new`.

**Reply by sawcce** [reply]
Tysm for your help! I guess I would have to implement my own require function if I wanted to import other bytecodes right?

**Comment by sawcce** [comment]
While I'm at it since it seems like the dump function is only usable with the LuaJIT feature I can't use luau with it right?

**Reply by khvzak** [reply]
There is a `Compiler` for Luau that can be used to compile source code. It's missing from the beta.1 documentation but present and works.
I'll release tomorrow beta.2 that will include fixed docs as well

**Reply by khvzak** [reply]
I pushed a new version which has `Compiler` struct for Luau: https://docs.rs/mlua/0.8.0-beta.2/mlua/struct.Compiler.html

**Reply by sawcce** [reply]
> I pushed a new version which has `Compiler` struct for Luau: https://docs.rs/mlua/0.8.0-beta.2/mlua/struct.Compiler.html
> 

Thanks I tried it out and it works great 👍
